### PR TITLE
ESLint for karate js - and other accommodating changes

### PR DESCRIPTION
**Pull request by pl-nhs**
## Summary

To accommodate the more recent packages for linting, there have been updates to the version of node we use and the npm packages. I took the opportunity to bump up the version of Node we use to lts/iron. The main impact was on the node sandbox, but simply updating the npm libraries used there seems to have been all that was needed. The docker container for this sandbox has also been updated to use the newer version of node js.

**Karate sandbox**

- The sandbox .js files have been linted according to the eslint standard rules, and I've added some documentation for the functions

**Github workflow**

- Update to the version of Node we install (and we're using different action to simplify the code too)
- I've removed the starting of the node sandbox from the github workflow, since this doesn't seem to be necessary

**Makefile**

- I changed the lint command to standardise the commands - it just calls commands from package.json now for the different linting routines

**Main README**

- Changed the triple quote comments to bash comments and removed the dollar signs from each command line
- Updated the parts relating to node version, added a bit about VSC extensions

**Karate README**

- Again, added a section about VSC extensions

**package.json**

- defined scripts for different linting routines, in line with what we have in the Makefile

**Node sandbox**

- Upgraded version of node to lts/iron, updated docker container and npm package versions

**Git pre-commit script**

- Simplified the lint command - we just need to call the `make lint` command here

**Pytest sandbox tests**

- Added a skipif tag so we don't try to run those _ping and _status checks locally (since they will fail)
- Tweaked code slightly to make the Python grammarians happier

## Reviews Required
* [x] Dev
* [ ] Test
* [ ] Tech Author
* [ ] Product Owner

## Review Checklist
:information_source: This section is to be filled in by the **reviewer**.

* [ ] I have reviewed the changes in this PR and they fill all or part of the acceptance criteria of the ticket, and the code is in a mergeable state.
* [ ] If there were infrastructure, operational, or build changes, I have made sure there is sufficient evidence that the changes will work.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -132,31 +137,27 @@ There are `make` commands that alias some of this functionality:
...
-```
+```bash
 make test-sandbox
 ```
-If a lot of the tests fail and the sandbox crashes with 
-```
-Cannot read properties of null (reading 'statusCode')
-```  
-you may not be using the correct version of node. Check the version being used by that instance of the terminal and run re-run '`nvm use lts/fermium`' if required.
+If a lot of the tests fail and the sandbox crashes with  the error `Cannot read properties of null (reading 'statusCode')`, you may not be using the correct version of node. Check the version being used by that instance of the terminal and run re-run '`nvm use lts/iron`' if required.
```
Comment (ehsanhussain-nhsd) [inline README.md:153]: Double space between with and the

File: karate-tests/src/test/java/mocks/sandbox/get-patient-search.js [hunk 2]
```diff
@@ -1,119 +1,122 @@
...
-            {
-                "fullUrl": "https://api.service.nhs.uk/personal-demographics/FHIR/R4/Patient/9000000009",
-                "search": {"score": score},
-                "resource": SEARCH_PATIENT_9000000009
-            }
-        ]
-    }
+
+/* Karate objects */
+/* global request, response */
```
Comment (ehsanhussain-nhsd) [inline karate-tests/src/test/java/mocks/sandbox/get-patient-search.js:6]: Are these comments still required?

File: karate-tests/src/test/java/mocks/sandbox/patch-patient.js [hunk 3]
```diff
@@ -1,146 +1,154 @@
-function buildResponseHeaders(request, patient) {
-    return {
-        'content-type': 'application/fhir+json',
-        'etag': `W/"${patient.meta.versionId}"`,
-        'x-request-id': request.header('x-request-id'),
-        'x-correlation-id': request.header('x-correlation-id')
-    }
+/* Karate objects */
```
Comment (ehsanhussain-nhsd) [inline karate-tests/src/test/java/mocks/sandbox/patch-patient.js:1]: Again, are these comments still needed?

File: karate-tests/src/test/java/mocks/sandbox/patch-patient.js [hunk 4]
```diff
@@ -1,146 +1,154 @@
...
-function setInvalidUpdateError(request, diagnostics) {
-    let body = context.read('classpath:mocks/stubs/errorResponses/INVALID_UPDATE.json')
-    body.issue[0].diagnostics = diagnostics
-    response.headers = basicResponseHeaders(request)
-    response.body = body
-    response.status = 400
-    return false
+function setInvalidUpdateError (request, diagnostics) {
+  const body = context.read('classpath:mocks/stubs/errorResponses/INVALID_UPDATE.json')
+  body.issue[0].diagnostics = diagnostics
```
Comment (ehsanhussain-nhsd) [inline karate-tests/src/test/java/mocks/sandbox/patch-patient.js:29]: `body` is defined as a const and we're changing a property within it. Should this be left as a `let` instead?

File: karate-tests/src/test/java/mocks/sandbox/patch-patient.js [hunk 5]
```diff
@@ -1,146 +1,154 @@
...
 
-function setPreconditionFailedError(request, diagnostics) {
-    let body = context.read('classpath:mocks/stubs/errorResponses/PRECONDITION_FAILED.json')
-    body.issue[0].diagnostics = diagnostics
-    response.headers = basicResponseHeaders(request)
-    response.body = body
-    response.status = 412
+function setPreconditionFailedError (request, diagnostics) {
+  const body = context.read('classpath:mocks/stubs/errorResponses/PRECONDITION_FAILED.json')
+  body.issue[0].diagnostics = diagnostics
```
Comment (ehsanhussain-nhsd) [inline karate-tests/src/test/java/mocks/sandbox/patch-patient.js:38]: Same here, should it be left as a `let`?

File: karate-tests/src/test/java/mocks/sandbox/stubs.js [hunk 6]
```diff
@@ -1,31 +1,38 @@
 /*
     Stubbed responses used by the different mocks
 */
+
+/* eslint-disable no-unused-vars */
+/* global context, session */
```
Comment (ehsanhussain-nhsd) [inline karate-tests/src/test/java/mocks/sandbox/stubs.js:6]: Are comments needed?

File: karate-tests/src/test/java/mocks/sandbox/supporting-functions.js [hunk 7]
```diff
@@ -1,102 +1,150 @@
...
-    response.status = 400
+/**
+ * Sets an invalid value error response for a given field and value.
+ * @param {string} field - The header field name.
+ * @param {string} value - The invalid value.
+ * @param {object} request - The request object.
+ */
+function setInvalidValueError (field, value, request) {
+  const body = context.read('classpath:mocks/stubs/errorResponses/INVALID_VALUE.json')
+  body.issue[0].diagnostics = `Invalid value - '${value}' in header '${field}'`
```
Comment (ehsanhussain-nhsd) [inline karate-tests/src/test/java/mocks/sandbox/supporting-functions.js:30]: Modifying const. Leave as `let`?

File: karate-tests/src/test/java/mocks/sandbox/supporting-functions.js [hunk 8]
```diff
@@ -1,102 +1,150 @@
...
+  response.status = 400
 }
 
-function setMissingValueError(diagnostics) {
-    let body = context.read('classpath:mocks/stubs/errorResponses/MISSING_VALUE.json')
-    body["issue"][0]["diagnostics"] = diagnostics
-    response.body = body
-    response.status = 400
+function setMissingValueError (diagnostics) {
+  const body = context.read('classpath:mocks/stubs/errorResponses/MISSING_VALUE.json')
```
Comment (ehsanhussain-nhsd) [inline karate-tests/src/test/java/mocks/sandbox/supporting-functions.js:37]: Modifying const. Leave as `let`?

File: karate-tests/src/test/java/mocks/sandbox/get-patient-search.js (hunk 2 above)
Comment (pl-nhs) [reply karate-tests/src/test/java/mocks/sandbox/get-patient-search.js:6]: Yes these comments are intentionally here - they're commands for esint to understand that these are globals

File: karate-tests/src/test/java/mocks/sandbox/supporting-functions.js [hunk 9]
```diff
@@ -1,102 +1,150 @@
...
+/**
+ * Sets an INVALID_UPDATE error response for the given request and diagnostics.
+ *
+ * @param {Object} request - The request object.
+ * @param {Array} diagnostics - The diagnostics to be set in the error response.
+ */
+// eslint-disable-next-line no-unused-vars
+function setInvalidUpdateError (request, diagnostics) {
+  const body = context.read('classpath:mocks/stubs/errorResponses/INVALID_UPDATE.json')
+  body.issue[0].diagnostics = diagnostics
```
Comment (ehsanhussain-nhsd) [inline karate-tests/src/test/java/mocks/sandbox/supporting-functions.js:65]: Modifying const

File: karate-tests/src/test/java/mocks/sandbox/patch-patient.js (hunk 4 above)
Comment (ehsanhussain-nhsd) [reply karate-tests/src/test/java/mocks/sandbox/patch-patient.js:29]: Objects within a const can be altered. Will remove similar comments

**Review by ehsanhussain-nhsd — APPROVED** [review]

**Pull request merged**